.from_bytes() implementation for long MPZ ints

pfalcon opened the pull request:
This is WIP for the discussion in https://github.com/micropython/micropython/issues/682#issuecomment-270852671 . The usual problem here is to make it work with all configuration options for long ints that we have, and that update tests for that. That takes me extended time, so I post this WIP for now (which works well for unix port and passes tests).

Commit:
py/mpz: Implement mpz_set_from_bytes() as a foundation for int.from_bytes().

Commit:
py/objint: Delegate from_bytes() to mp_obj_int_from_bytes_impl().

An implementation of int.from_bytes() is now delegated to function
mp_obj_int_from_bytes_impl() of particular objint implementation. A
version for objint_mpz is provided.

Commit:
tests: Add test for int.from_bytes() for arbitrary-precision integer.

This test works only for MICROPY_LONGINT_IMPL == MICROPY_LONGINT_IMPL_MPZ
and needs a way of skipping in other cases.

pfalcon commented:
@nickovs: FYI.

dpgeorge commented on py/objint.c:
```diff
@@ -374,26 +374,27 @@ mp_obj_t mp_obj_int_binary_op_extra_cases(mp_uint_t op, mp_obj_t lhs_in, mp_obj_
     mp_get_buffer_raise(args[1], &bufinfo, MP_BUFFER_READ);
 
-    // convert the bytes to an integer
+    #if MICROPY_LONGINT_IMPL != MICROPY_LONGINT_IMPL_NONE
```
It would be desirable to use the original code path (ie convert to small-int) if the value fits in a small integer.  To keep it simple one could just do something like `if (MP_SMALL_INT_FITS(1 << (bufinfo.len * 8 -1))) ...`.

pfalcon replied:
I of course thought about that, but checking exact conditions for that is boring, so as you point, only heuristic conditions would fit. And then, endian="big" was not implemented for that case. Ok, now I implemented "big" case, that's good. But I still don't think it's worth to spend bytes on this heuristic optimization - not before it was shown to be useful at lease. We have optimizations found to be useful from practice (e.g. https://github.com/micropython/micropython/issues/2719), it's better to save bytes for those.

dpgeorge replied:
> But I still don't think it's worth to spend bytes on this heuristic optimization - not before it was shown to be useful at lease

A case where it would be useful is instead of `struct.unpack('I', buf)` which creates a tuple to return the single value.  Instead one could use `int.from_bytes` and it could work without allocating on the heap.  See eg end of esp8266/modules/ds18x20.py which does ad-hoc bit shifting to unpack a signed 2-byte integer, in order to avoid allocating memory.

pfalcon replied:
It won't help with `struct.unpack('I', buf)`, only with `struct.unpack('H', buf)`, which is 2 bytes, which is indeed can be easily replaced with shift-or. I will add that optimization later, unless you finally agree it doesn't add much useful ;-).

dpgeorge replied:
I'm getting confused here, and just realised that the title of this PR is wrong, should be from_bytes not to_bytes.

You are right that "I" wont use the small-int path with the simple check for length of incoming buffer.  But the more useful case is 'h', ie a signed 2-byte conversion, because that requires more than a shift-or in Python to implement.

My main concern is that this PR is a breaking-change, in the sense that someone who already uses int.from_bytes with the heap locked (eg in an irq) will find that their code stops working after this is merged.

pfalcon replied:
Title is fixed, sorry. Signed conversion for .from_bytes() is again not implemented. But sounds good, I'll implement what you suggest.

dpgeorge reviewed (commented):
> The usual problem here is to make it work with all configuration options for long ints that we have

What you have here is a good start for the most used case of mpz.  Other implementations can follow later, but at least put a dummy implementation of `mp_obj_int_from_bytes_impl` for non-mpz cases.

pfalcon commented on py/mpz.c:
```diff
@@ -909,6 +909,36 @@ mp_uint_t mpz_set_from_str(mpz_t *z, const char *str, mp_uint_t len, bool neg, m
     return cur - str;
 }
 
+void mpz_set_from_bytes(mpz_t *z, bool big_endian, mp_uint_t len, byte *buf) {
```
Done

pfalcon commented on py/objint_mpz.c:
```diff
@@ -107,6 +107,12 @@ char *mp_obj_int_formatted_impl(char **buf, size_t *buf_size, size_t *fmt_size,
     return str;
 }
 
+mp_obj_t mp_obj_int_from_bytes_impl(bool big_endian, size_t len, byte *buf) {
```
Done

pfalcon commented on py/mpz.c:
```diff
@@ -909,6 +909,36 @@ mp_uint_t mpz_set_from_str(mpz_t *z, const char *str, mp_uint_t len, bool neg, m
+
+    mpz_dig_t d = 0;
+    int num_bits = 0;
+    z->len = 0;
```
Done

Commit:
py/objint: from_bytes: Support byteorder="big" for MICROPY_LONGINT_IMPL_NONE.

dpgeorge commented on py/objint.h:
```diff
@@ -59,6 +59,7 @@ char *mp_obj_int_formatted(char **buf, size_t *buf_size, size_t *fmt_size, mp_co
 char *mp_obj_int_formatted_impl(char **buf, size_t *buf_size, size_t *fmt_size, mp_const_obj_t self_in,
                                 int base, const char *prefix, char base_char, char comma);
 mp_int_t mp_obj_int_hash(mp_obj_t self_in);
+mp_obj_t mp_obj_int_from_bytes_impl(bool big_endian, size_t len, byte *buf);
```
Did you make this const?

pfalcon replied:
Now done.

Commit:
py/objint_longlong: Add stub for mp_obj_int_from_bytes_impl().

To be implemented later.

Commit:
py/objint: from_bytes(): If result fits in small int, use that.

Commit:
tests/heapalloc_int_from_bytes: Test that int.from_bytes() can work w/o alloc.

For a small number of bytes, it's expected to return a small int without
allocation.

pfalcon commented:
Ok, this is fully implemented, I'm flushing it to master.

pfalcon commented:
Squashed a bit and merged.

dpgeorge commented:
Thanks!

nickovs commented:
For the unix port, when I test this on Linux it works correctly but when I test this on MacOS it does not work. After a `make clean; make axtls; make` on Linux I see:
```
nicko@prolapse:~/micropython/unix$ git show --oneline -s
b32a38e esp8266: Factor out common linker code to esp8266_common.ld.
nicko@prolapse:~/micropython/unix$ ./micropython -c 'print(int.from_bytes(b"\xff"*64, "little"))'
13407807929942597099574024998205846127479365820592393377723561443721764030073546976801874298166903427690031858186486050853753882811946569946433649006084095
```
When I do the same on MacOS I get:
```
Nickos-MBP-84:unix nicko$ git show --oneline -s
b32a38e esp8266: Factor out common linker code to esp8266_common.ld.
Nickos-MBP-84:unix nicko$ ./micropython -c 'print(int.from_bytes(b"\xff"*64, "little"))'
18446744073709551615
```
I'm not quite sure what's going on here but something is clearly not right.

dpgeorge commented:
> I'm not quite sure what's going on here but something is clearly not right.

Thanks for the report!  Should be fixed by eaa77455c37cf18670d181556e6d17e0b01c2267

nickovs commented:
Yes, that seems to fix it.
